### PR TITLE
Qmaps 507 metrics

### PR DESCRIPTION
**Pull request by Julien-laville**
Add common telemetry event list
Create front telemetry class
Add telemetry to
-poi
-fav
-share
-service

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -991,7 +991,7 @@
     },
     "@sinonjs/formatio": {
       "version": "2.0.0",
-      "resolved": "http://registry.npmjs.org/@sinonjs/formatio/-/formatio-2.0.0.tgz",
+      "resolved": "https://registry.npmjs.org/@sinonjs/formatio/-/formatio-2.0.0.tgz",
```
Comment (amatissart) [inline package-lock.json]: npm warns with low severity vulnerabilities. To fix with `npm audit fix`

File: src/modals/share.js [hunk 2]
```diff
@@ -31,7 +32,12 @@ export default class Share {
     this.modal.close()
   }
 
-  openPopup () {
+  openPopup (targetName) {
+    if(targetName === 'facebook') {
+      Telemetry.add(Telemetry.SHARE_FACEBOOK)
```
Comment (amatissart) [inline src/modals/share.js]: These events are triggered on each share modal opening and closing, and on app initialization :upside_down_face:

File: src/mapbox/extended_geolocate_control.js [hunk 3]
```diff
@@ -21,7 +30,10 @@ class ExtendedGeolocateControl extends GeolocateControl {
 
   trigger() {
     window.navigator.permissions && window.navigator.permissions.query({ name: 'geolocation' }).then(p => {
-      if (p.state === 'prompt') {
+      if(p.state === geolocationPermissions.GRANTED) {
+        Telemetry.add(Telemetry.LOCALISE_START)
```
Comment (amatissart) [inline src/mapbox/extended_geolocate_control.js]: This event will not be triggered when prompt is used (on the first visit for example)

File: src/libs/telemetry.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import Ajax from './ajax'
+import nconf from '@qwant/nconf-getter'
+import telemetryModule from 'telemetry'
+
+const telemetry = nconf.get().telemetry
+
+export default class Telemetry {
+  constructor() {}
+
+  static add(event) {
```
Comment (amatissart) [inline src/libs/telemetry.js]: Maybe we could find a more explicit name for this function ? `trigger` ?

File: src/libs/telemetry.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    return Telemetry.send(event)
+  }
+
+  static async send(event) {
+    let data = {type : event}
+    let telemetryUrl = `${telemetry.eventEndpoint}`
```
Comment (amatissart) [inline src/libs/telemetry.js]: What about `baseUrl` ? I think we need to add it as a prefix here.

File: config/default_config.yml [hunk 6]
```diff
@@ -47,3 +47,7 @@ mapPlugins:
 
 statics:
   maxAge: 0  # string accepted by the ms module, or milliseconds
+
+telemetry:
+  logsEndpoint: '/logs'
+  eventEndpoint: '/events'
```
[the remark this answers is not in the document]
Comment (Julien-laville) [reply config/default_config.yml]: i keep telemtry enbled inside config & move endPoints to telemetry file as const

File: package-lock.json (hunk 1 above)
Comment (Julien-laville) [reply package-lock.json]: fixed

File: src/modals/share.js (hunk 2 above)
Comment (Julien-laville) [reply src/modals/share.js]: i dont reproduce on app init @amatissart

File: src/panel/poi_panel.js [hunk 7]
```diff
@@ -68,6 +69,7 @@ PoiPanel.prototype.close = async function() {
 }
 
 PoiPanel.prototype.restorePoi = async function (id) {
+  Telemetry.add(Telemetry.POI_RESTORE)
```
Comment (amatissart) [inline src/panel/poi_panel.js]: `poi_restore` event is undefined.  
This causes a request without body to backend, that returns 400. 
I think `Telemetry.add` should validate the event, and log invalid events instead of triggering an invalid request.

File: src/mapbox/extended_geolocate_control.js [hunk 8]
```diff
@@ -1,12 +1,24 @@
...
+  GRANTED : 'granted',
+  DENIED : 'denied'
+}
+
 class ExtendedGeolocateControl extends GeolocateControl {
   constructor(options, container) {
     super(options)
     this._container = container
+    this.on('trackuserlocationstart', () => {
+      Telemetry.add(Telemetry.LOCALISE_TRIGGER)
```
Comment (amatissart) [inline src/mapbox/extended_geolocate_control.js]: I am not sure this is a very relevant metric. If location is refused or blocked for some reason, we'd still count each tap on the button.  
Why not triggering this event on the first click only ?

Comment (Julien-laville) [reply src/mapbox/extended_geolocate_control.js]: i propose a new telemetry method to complete add() : addOnce().

File: src/panel/poi_panel.js (hunk 7 above)
Comment (Julien-laville) [reply src/panel/poi_panel.js]: good idea

File: config/default_config.yml [hunk 9]
```diff
@@ -47,3 +47,5 @@ mapPlugins:
 
 statics:
   maxAge: 0  # string accepted by the ms module, or milliseconds
+
+telemetryEnabled: true
```
[the remark this answers is not in the document]
Comment (Julien-laville) [reply config/default_config.yml]: i prefer keep it simple for now, and update to nested the day we have multiple configuration

File: src/libs/telemetry.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+  static add(event) {
+    return Telemetry.send(event)
+  }
+
+  static async send(event) {
+    if(telemetryEnabled) {
+      let data = {type : event}
+      let telemetryUrl = `${system.baseUrl}${telemtryEventUrl}`
+      console.log(telemetryUrl)
```
[the remark this answers is not in the document]
Comment (Julien-laville) [reply src/libs/telemetry.js]: done

File: src/adapters/autocomplete.js [hunk 11]
```diff
@@ -36,7 +36,7 @@ function SearchInput(tagSelector) {
       let isAbort = false
       let promise = new Promise((resolve, reject) => {
         /* 'bbox' is currently not used by the geocoder, it' will be used for the telemetry. */
-        this.suggestPromise = ajax.query(geocoderUrl, {q: term})
+        this.suggestPromise = ajax.get(geocoderUrl, {q: term})
```
Comment (amatissart) [inline src/adapters/autocomplete.js]: Another `ajax.query` is to change on line 95 ;)

File: config/default_config.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (amatissart) [reply config/default_config.yml]: I will insist on that one :)  
We will need other settings about logging of telemetry events very soon.  
So I'd like to keep backward compatible configuration on the different environments.
